pkg/monitor/disruption_backend_sampler.go

```diff
@@ -0,0 +1,349 @@
+		if err != nil {
+			hostErr = err
+			return
+		}
+		for _, ingress := range route.Status.Ingress {
+			if len(ingress.Host) > 0 {
+				b.host = fmt.Sprintf("https://%s", ingress.Host)
+				break
+			}
+		}
+	})
+	if hostErr != nil {
+		return "", hostErr
+	}
+	if len(b.host) == 0 {
+		return "", fmt.Errorf("missing URL")
+	}
+	return b.host + b.path, nil
+}
+
+func (b *BackendSampler) getTLSConfig() *tls.Config {
+	if b.tlsConfig == nil {
+		return &tls.Config{InsecureSkipVerify: true}
+	}
+	return b.tlsConfig
+}
+
+func (b *BackendSampler) wrapWithAuth(rt http.RoundTripper) (http.RoundTripper, error) {
+	if len(b.bearerToken) == 0 && len(b.bearerTokenFile) == 0 {
+		return rt, nil
+	}
+
+	if b.tlsConfig == nil {
+		return nil, fmt.Errorf("WithTLSConfig is required if you are providing a token")
+	}
+
+	return transport.NewBearerAuthWithRefreshRoundTripper(b.bearerToken, b.bearerTokenFile, rt)
+}
+
+func (b *BackendSampler) GetHTTPClient() (*http.Client, error) {
+	var httpTransport *http.Transport
+	switch b.GetConnectionType() {
+	case NewConnectionType:
+		httpTransport = &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout:   b.getTimeout(),
+				KeepAlive: -1, // this looks unnecessary to me, but it was set in other code.
+			}).Dial,
+			TLSClientConfig:     b.getTLSConfig(),
+			TLSHandshakeTimeout: b.getTimeout(),
+			DisableKeepAlives:   true, // this prevents connections from being reused
+			IdleConnTimeout:     b.getTimeout(),
+		}
+
+	case ReusedConnectionType:
+		httpTransport = &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: b.getTimeout(),
+			}).Dial,
+			TLSClientConfig:     b.getTLSConfig(),
+			TLSHandshakeTimeout: b.getTimeout(),
+			IdleConnTimeout:     b.getTimeout(),
+		}
+
+	default:
+		return nil, fmt.Errorf("unrecognized connection type")
+	}
+
+	roundTripper, err := b.wrapWithAuth(http.RoundTripper(httpTransport))
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient := &http.Client{
+		Transport: roundTripper,
+	}
+
+	return httpClient, nil
+}
+
+// StartEndpointMonitoring sets up a client for the given BackendSampler and starts a
+// new sampler for the given monitor that uses the client to monitor
+// connectivity to the BackendSampler and reports any observed disruption.
+//
+// If disableConnectionReuse is false, the client reuses connections and detects
+// abrupt breaks in connectivity.  If disableConnectionReuse is true, the client
+// instead creates fresh connections so that it detects failures to establish
+// connections.
+func (b *BackendSampler) StartEndpointMonitoring(ctx context.Context, m *Monitor) error {
+	httpClient, err := b.GetHTTPClient()
+	if err != nil {
+		return err
+	}
+
+	url, err := b.GetURL()
+	if err != nil {
+		return err
+	}
+
+	go NewSampler(m, time.Second, func(previous bool) (condition *monitorapi.Condition, next bool) {
+		resp, getErr := httpClient.Get(url)
```

Comment:
Are we missing `defer resp.Body.Close()` ?

If we use http/1.x on anything, not closing the Body will break connection reuse.  Even if not, it's probably a good idea to close it anyway.

https://pkg.go.dev/net/http#Response:

```go
	// It is the caller's responsibility to
	// close Body. The default HTTP client's Transport may not
	// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
	// not read to completion and closed.
```